nx/lib/nx/defn/grad.ex

```diff
@@ -476,6 +476,41 @@ defmodule Nx.Defn.Grad do
     grad_pairs(pairs, g, cache)
   end
 
+  defp grad(:qr, [{q, r}, input, _opts], _ans, g, cache) do
+    r_inv = Nx.LinAlg.invert(r)
+    c = Nx.transpose(q) |> Nx.dot(g) |> Nx.dot(r_inv)
+
+    # matrix where every element above the main diagonal is 2
+    # and every element on the main diagonal is 1
+    e =
+      Nx.less_equal(Nx.iota(c, axis: 0), Nx.iota(c, axis: 1))
+      |> Nx.multiply(2)
+      |> Nx.subtract(Nx.eye(c))
+
+    dr = c |> Nx.add(Nx.transpose(c)) |> Nx.divide(2) |> Nx.multiply(e) |> Nx.dot(r)
+    dq = g |> Nx.subtract(Nx.dot(q, dr)) |> Nx.dot(r_inv)
+
+    # The code below yields the following thensor for the test at  test/nx/defn/grad_test.exs:1658
+    # #Nx.Tensor<\n  f32[2][2]\n  [\n    [1.9999998807907104, -0.33333325386047363],\n    [-0.3333333134651184, -0.9999999403953552]\n  ]\n>
+    # While the code above yields
+    # #Nx.Tensor<\n  f32[2][2]\n  [\n    [2.0, -0.3333331346511841],\n    [-0.33333325386047363, -0.9999999403953552]\n  ]\n>
+
+    # The code below was translated from jax, while the code above was translated from
+    # https://arxiv.org/pdf/2009.10071.pdf, page 5
```

Comment:
@josevalim I've just found this, which may tell us that we're on the right track but I'm missing something.

Basically, two different codes which are based on two different sources are yielding equivalent tensors (but not the expected output when compared to Jax).

The only thing I could think of now is that there is some fundamental difference between `grad/5` and `jvp/3` which would make sense of this.

Any thoughts?